codegen/templates/contrib/sets/sets.gotmpl

```diff
@@ -0,0 +1,114 @@
+package {{.Version}}
+
+import (
+    {{- if $.CustomTypesImportPath }}
+
+    . "{{ $.CustomTypesImportPath }}"
+    {{- end}}
+)
+
+// k8s resources are uniquely identified by their name and namespace
+func key(objectMeta v1.ObjectMeta) string {
+    return objectMeta.GetName() + "." + objectMeta.GetNamespace()
+}
+
+{{- range $resource := $.Resources }}
+
+{{- $kindLowerCamel := lower_camel $resource.Kind }}
+
+type {{ $resource.Kind }}Set interface {
+    Keys() sets.String
+    List() []*{{ $resource.Kind }}
+    Map() map[string]*{{ $resource.Kind }}
+    Insert({{ $kindLowerCamel }} ...*{{ $resource.Kind }})
+    Equal({{ $kindLowerCamel }}Set {{ $resource.Kind }}Set) bool
+    Has({{ $kindLowerCamel }} *{{ $resource.Kind }}) bool
+    Delete({{ $kindLowerCamel }} *{{ $resource.Kind }})
+    Union(set {{ $resource.Kind }}Set) {{ $resource.Kind }}Set
+    Difference(set {{ $resource.Kind }}Set) {{ $resource.Kind }}Set
+    Intersection(set {{ $resource.Kind }}Set) {{ $resource.Kind }}Set
+}
+
+type {{ $kindLowerCamel }}Set struct {
+    set     sets.String
+    mapping map[string]*{{ $resource.Kind }}
+}
+
+func New{{ $resource.Kind }}Set({{ $resource.Kind }}s ...*{{ $resource.Kind }}) {{ $resource.Kind }}Set {
+    set := sets.NewString()
+    mapping := map[string]*{{ $resource.Kind }}{}
+    for _, {{ $resource.Kind }} := range {{ $resource.Kind }}s {
+        key := key({{ $resource.Kind }}.ObjectMeta)
+        set.Insert(key)
+        mapping[key] = {{ $resource.Kind }}
+    }
+    return &{{ $kindLowerCamel }}Set{set: set, mapping: mapping}
+}
+
+func (s {{ $kindLowerCamel }}Set) Set() sets.String {
+    return s.set
```

Comment:
same comment here, this is a mutable internal datastructure, best practice is to copy it into a new thing before handing it out to callers